clang/lib/Sema/SemaDecl.cpp

```diff
@@ -11802,6 +11802,27 @@ static bool CheckMultiVersionFunction(Sema &S, FunctionDecl *NewFD,
                                          OldDecl, Previous);
 }
 
+static void CheckFunctionDeclarationAttributesUsage(Sema &S,
+                                                    FunctionDecl *NewFD) {
+  bool IsPure = NewFD->hasAttr<PureAttr>();
+  bool IsConst = NewFD->hasAttr<ConstAttr>();
+
+  if (IsPure && IsConst) {
+    S.Diag(NewFD->getLocation(), diag::warn_const_attr_with_pure_attr);
+    NewFD->dropAttr<PureAttr>();
+  }
+  if (IsPure || IsConst) {
```

Comment:
```suggestion
  if (!IsPure && !IsConst)
     return;
   ...
```

nit: we prefer short indentation.

Reply:
Early return is good, but i think it may be customization point, where different attributes/contract checks will be in future, so early return will cause more changes in future

Reply:
Then we should handle different attributes in different places.